File: src/main/java/com/kotakotik/chaoscraftfabric/utils/CCReflections.java
package com.kotakotik.chaoscraftfabric.utils;

import com.kotakotik.chaoscraftfabric.ChaoscraftFabric;
import org.reflections.Reflections;
import org.reflections.scanners.SubTypesScanner;
import org.reflections.scanners.TypeAnnotationsScanner;

import java.lang.annotation.Annotation;
import java.util.Set;

import static org.reflections.ReflectionUtils.forNames;
import static org.reflections.ReflectionUtils.withAnnotation;
import static org.reflections.util.Utils.filter;
import static org.reflections.util.Utils.names;

public abstract class CCReflections {
    public static final Reflections reflections = new Reflections(ChaoscraftFabric.class.getPackage().getName());

    public static  <T> Set<Class<? extends T>> getSubTypesOf(Class<T> type) {
        return reflections.getSubTypesOf(type);
    }

    public static Set<Class<?>> getTypesAnnotatedWith(final Annotation annotation) {
        return reflections.getTypesAnnotatedWith(annotation, true);
    }

    public static Set<Class<?>> getTypesAnnotatedWith(final Class<? extends Annotation> annotation) {
        return reflections.getTypesAnnotatedWith(annotation, true);
    }
}

File: src/main/kotlin/com/kotakotik/chaoscraft/utils/CCReflections.kt
package com.kotakotik.chaoscraft.utils

import com.kotakotik.chaoscraft.ChaoscraftKotlin
import org.reflections.Reflections

object CCReflections {
    val reflections: Reflections = Reflections(ChaoscraftKotlin::class.java.getPackage().getName())
    fun <T> getSubTypesOf(type: Class<T>?): Set<Class<out T>> {
        return reflections.getSubTypesOf(type)
    }

    fun getTypesAnnotatedWith(annotation: Annotation?): Set<Class<*>> {
        return reflections.getTypesAnnotatedWith(annotation, true)
    }

    fun getTypesAnnotatedWith(annotation: Class<out Annotation?>?): Set<Class<*>> {
        return reflections.getTypesAnnotatedWith(annotation, true)
    }
}